worker/src/ocall_bridge/on_chain_ocall_impl.rs

```diff
@@ -0,0 +1,153 @@
+    Copyright (C) 2017-2019 Baidu, Inc. All Rights Reserved.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+use crate::node_api_factory::NodeApiFactory;
+use crate::ocall_bridge::bridge_api::{OCallBridgeError, OCallBridgeResult, WorkerOnChainOCall};
+use crate::sync_block_gossiper::SyncBlockGossiper;
+use crate::utils::hex_encode;
+use crate::{WorkerRequest, WorkerResponse};
+use codec::{Decode, Encode};
+use log::*;
+use sp_core::storage::StorageKey;
+use std::sync::mpsc::channel;
+use std::sync::Arc;
+use std::vec::Vec;
+use substrate_api_client::XtStatus;
+use substratee_worker_primitives::block::SignedBlock as SignedSidechainBlock;
+
+pub struct WorkerOnChainOCallImpl<F, S>
+where
+    F: NodeApiFactory,
+    S: SyncBlockGossiper,
+{
+    node_api_factory: Arc<F>,
+    block_gossiper: Arc<S>,
+}
+
+impl<F, S> WorkerOnChainOCallImpl<F, S>
+where
+    F: NodeApiFactory,
+    S: SyncBlockGossiper,
+{
+    pub fn new(node_api_factory: Arc<F>, block_gossiper: Arc<S>) -> Self {
+        WorkerOnChainOCallImpl {
+            node_api_factory,
+            block_gossiper,
+        }
+    }
+}
+
+impl<F, S> WorkerOnChainOCall for WorkerOnChainOCallImpl<F, S>
+where
+    F: NodeApiFactory,
+    S: SyncBlockGossiper,
+{
+    fn worker_request(&self, request: Vec<u8>) -> OCallBridgeResult<Vec<u8>> {
+        debug!("    Entering ocall_worker_request");
+
+        let requests: Vec<WorkerRequest> = Decode::decode(&mut request.as_slice()).unwrap();
+        if requests.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let api = self.node_api_factory.create_api();
+
+        let resp: Vec<WorkerResponse<Vec<u8>>> = requests
+            .into_iter()
+            .map(|req| match req {
+                WorkerRequest::ChainStorage(key, hash) => WorkerResponse::ChainStorage(
+                    key.clone(),
+                    api.get_opaque_storage_by_key_hash(StorageKey(key.clone()), hash)
+                        .unwrap(),
+                    api.get_storage_proof_by_keys(vec![StorageKey(key)], hash)
+                        .unwrap()
+                        .map(|read_proof| {
+                            read_proof.proof.into_iter().map(|bytes| bytes.0).collect()
+                        }),
+                ),
+            })
+            .collect();
+
+        let encoded_response: Vec<u8> = resp.encode();
+
+        Ok(encoded_response)
+    }
+
+    fn send_block_and_confirmation(
```

Comment:
In case we need to separate the sending of block and confirmations, would this be a huge hassle to do this later on?
See https://github.com/integritee-network/worker/issues/301

Reply:
I'd say with this refactoring, that separation will be quite a bit easier to do since we have an object with state that both methods can access.

Reply:
Perfect, then we can leave this issue for the future.